typing/typecore.ml

```diff
@@ -2542,72 +2542,76 @@ let generalize_and_check_univars env kind exp ty_expected vars =
-                check e
-            | Texp_apply _ | Texp_send _ | Texp_new _ | Texp_letop _ ->
-                Location.prerr_warning exp_loc
-                  Warnings.Ignored_partial_application
-          end
-        in
-        check exp
-    | Tvar _ ->
-        if delay then add_delayed_check (fun () -> f false)
+    | Tconstr (p, _, _)  when Path.same p Predef.path_unit -> ()
+    | Tvar _ -> ()
     | _ ->
-        check_statement ()
+        if statement then
+          let rec loop {exp_loc; exp_desc; exp_extra; _} =
+            match exp_desc with
+            | Texp_let (_, _, e)
+            | Texp_sequence (_, e)
+            | Texp_letexception (_, e)
+            | Texp_letmodule (_, _, _, _, e) ->
+                loop e
+            | _ ->
+                let loc =
+                  match List.find_opt (function
+                      | (Texp_constraint _, _, _) -> true
+                      | _ -> false) exp_extra
+                  with
+                  | Some (_, loc, _) -> loc
+                  | None -> exp_loc
+                in
+                Location.prerr_warning loc Warnings.Non_unit_statement
+          in
+          loop exp
   in
-  f true
+  match ty with
+  | Tarrow _ ->
+      let rec check {exp_desc; exp_loc; exp_extra; _} =
+        if List.exists (function
+            | (Texp_constraint _, _, _) -> true
+            | _ -> false) exp_extra then check_statement ()
+        else begin
+          match exp_desc with
+          | Texp_ident _ | Texp_constant _ | Texp_tuple _
+          | Texp_construct _ | Texp_variant _ | Texp_record _
+          | Texp_field _ | Texp_setfield _ | Texp_array _
+          | Texp_while _ | Texp_for _ | Texp_instvar _
+          | Texp_setinstvar _ | Texp_override _ | Texp_assert _
+          | Texp_lazy _ | Texp_object _ | Texp_pack _ | Texp_unreachable
+          | Texp_extension_constructor _ | Texp_ifthenelse (_, _, None)
+          | Texp_function _ ->
+              check_statement ()
+          | Texp_match (_, cases, _) ->
+              List.iter (fun {c_rhs; _} -> check c_rhs) cases
+          | Texp_try (e, cases) ->
+              check e; List.iter (fun {c_rhs; _} -> check c_rhs) cases
+          | Texp_ifthenelse (_, e1, Some e2) ->
+              check e1; check e2
+          | Texp_let (_, _, e) | Texp_sequence (_, e) | Texp_open (_, e)
+          | Texp_letexception (_, e) | Texp_letmodule (_, _, _, _, e) ->
+              check e
+          | Texp_apply _ | Texp_send _ | Texp_new _ | Texp_letop _ ->
+              Location.prerr_warning exp_loc
+                Warnings.Ignored_partial_application
+        end
+      in
+      check exp
+  | _ ->
+      check_statement ()
+
+let check_partial_application statement exp =
+  let ty = get_desc (expand_head exp.exp_env exp.exp_type) in
+  let doit () = check_partial_application statement exp in
+  match ty with
+  | Tvar _ ->
+      add_delayed_check doit
+  | _ ->
+      doit ()
```

Comment:
(Looking at the 'clarify delay logic' commit only) I like the clarification but I'm not fond of the shadowing here. I would pick two names (the first function could be `check_partial_application_now` for example), or keep the first definition inside the local `let doit ()` declaration.

Reply:
There is no more shadowing in the latest version of the code. Thanks!